staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -315,23 +320,99 @@ func strategicPatchObject(
 
 type applyPatcher struct {
 	*patcher
+
+	model proto.Schema
 }
 
-func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
-	versionedObjToUpdate, err := p.creater.New(p.kind)
+// TODO(apelisse): workflowId needs to be passed as a query
+// param/header, and a better defaulting needs to be defined too.
+const workflowId = "default"
+
+func (p *applyPatcher) convertCurrent(obj runtime.Object) (map[string]interface{}, error) {
+	vo, err := p.unsafeConvertor.ConvertToVersion(obj, p.kind.GroupVersion())
 	if err != nil {
 		return nil, err
 	}
-	if err := runtime.DecodeInto(p.codec, p.patchBytes, versionedObjToUpdate); err != nil {
-		return nil, err
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
+}
+
+func (p *applyPatcher) extractLastIntent(obj runtime.Object, workflow string) (map[string]interface{}, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get accessor: %v", err)
 	}
-	// For now, instead of performing a structured merge we will just attempt a PUT
-	// So we just convert the object back to unversioned (aka internal version).
-	unversionedObjToUpdate, err := p.toUnversioned(versionedObjToUpdate)
+	last := make(map[string]interface{})
+	if accessor.GetLastApplied()[workflow] != "" {
+		if err := json.Unmarshal([]byte(accessor.GetLastApplied()[workflow]), &last); err != nil {
+			return nil, fmt.Errorf("couldn't unmarshal last applied field: %v", err)
+		}
+	}
+	return last, nil
+}
+
+func (p *applyPatcher) getNewIntent() (map[string]interface{}, error) {
+	patch := make(map[string]interface{})
+	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
+		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
+	}
+	return patch, nil
+}
+
+func (p *applyPatcher) convertResult(result interface{}) (runtime.Object, error) {
+	voutput, err := p.creater.New(p.kind)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to create empty output object: %v", err)
 	}
-	return unversionedObjToUpdate, nil
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(result.(map[string]interface{}), voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert merge result back: %v", err)
+	}
+	p.defaulter.Default(voutput)
+
+	uoutput, err := p.toUnversioned(voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to unversioned: %v", err)
+	}
+
+	return uoutput, nil
+}
+
+func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
+	current, err := p.convertCurrent(currentObject)
```

Comment:
Maybe we can add a TODO about checking for and handling the case where currentObject and lastIntent are different versions.

Reply:
I'm not sure I agree. `convertCurrent` converts to the version of the end-point, which should be the versinon of the patch sent?